common/thrift_router.h

```diff
@@ -300,6 +288,7 @@ class ThriftRouter {
           ret = ReturnCode::NOT_FOUND;
           continue;
         }
+        createOrFixClientsFor(hosts_for_shard);
```

Comment:
Here we still do createOrFix for all hosts (hundreds of host in muse root) that host this shard, right?

Reply:
good catch! I didn't realize `filterByRoleAndSortByPreference` also returns full list of hosts. But i think even just getting rid of the hash set above is a good optimization for us. I applied the current patch to production and it shows ~10% cpu usage and a few milliseconds latency reduction.